api/types/access_list.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"time"
+
+	"github.com/gravitational/trace"
+
+	accesslistpb "github.com/gravitational/teleport/api/gen/proto/go/accesslist/v1"
+	"github.com/gravitational/teleport/api/types/header"
+	"github.com/gravitational/teleport/api/types/traits"
+)
+
+// FromAccessListV1 converts a v1 access list into an internal access list object.
+func FromAccessListV1(msg *accesslistpb.AccessList) (*AccessList, error) {
+	owners := make([]*AccessListOwner, len(msg.Spec.Owners))
+	for i, owner := range msg.Spec.Owners {
+		owners[i] = &AccessListOwner{
+			Name:        owner.Name,
+			Description: owner.Description,
+		}
+	}
+
+	members := make([]*AccessListMember, len(msg.Spec.Members))
+	for i, member := range msg.Spec.Members {
+		members[i] = &AccessListMember{
+			Name:    member.Name,
+			Joined:  member.Joined.AsTime(),
+			Expires: member.Expires.AsTime(),
+			Reason:  member.Reason,
+			AddedBy: member.AddedBy,
+		}
+	}
+
+	accessList, err := NewAccessList(header.FromMetadataV1(msg.Header.Metadata), &AccessListSpec{
+		Owners: owners,
+		Audit: &AccessListAudit{
+			Frequency: msg.Spec.Audit.Frequency.AsDuration(),
+		},
+		MembershipRequires: &AccessListRequires{
+			Roles:  msg.Spec.MembershipRequires.Roles,
+			Traits: traits.FromV1(msg.Spec.MembershipRequires.Traits),
+		},
+		OwnershipRequires: &AccessListRequires{
+			Roles:  msg.Spec.OwnershipRequires.Roles,
+			Traits: traits.FromV1(msg.Spec.OwnershipRequires.Traits),
+		},
+		Grants: &AccessListGrants{
+			Roles:  msg.Spec.Grants.Roles,
+			Traits: traits.FromV1(msg.Spec.Grants.Traits),
+		},
+		Members: members,
+	})
+
+	return accessList, trace.Wrap(err)
+}
+
+// AccessList describes the basic building block of access grants, which are
+// similar to access requests but for longer lived permissions that need to be
+// regularly audited.
+type AccessList struct {
+	// ResourceHeader is the common resource header for all resources.
+	*header.ResourceHeader
+
+	// Spec is the specification for the access list.
+	Spec *AccessListSpec `json:"spec" yaml:"spec"`
+}
+
+// AccessListSpec is the specification for an access list.
+type AccessListSpec struct {
+	// Description is a plaintext description of the access list.
+	Description string `json:"description" yaml:"description"`
+
+	// Owners is a list of owners of the access list.
+	Owners []*AccessListOwner `json:"owners" yaml:"owners"`
```

Comment:
Why not just a `[]AccessListOwner`?

Reply:
Same reason as https://github.com/gravitational/teleport/pull/28479#discussion_r1253371733. I'm happy to adjust if you feel strongly, though.

Reply:
Slices are fast due to locality. Slices of pointers to objects that live all over the place no longer has this advantage.

When it's the same amount of work either way, I don't see any reason to do the less efficient option.